lib/libc/stdlib/malloc/mrs/mrs.3

```diff
@@ -0,0 +1,231 @@
+.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+.\" SUCH DAMAGE.
+.\"
+.Dd October 18, 2023
+.Dt MRS 3
+.Os
+.Sh NAME
+.Nm MRS
+.Nd Malloc Revocation Shim
+.Sh LIBRARY
+.Lb libc
+.Sh SYNOPSIS
+.In stdlib.h
+.Ft void *
+.Fn malloc "size_t size"
+.Ft void *
+.Fn calloc "size_t number, size_t size"
+.Ft int
+.Fn posix_memalign "void **ptr, size_t alignment, size_t size"
+.Ft void *
+.Fn aligned_alloc "size_t alignment, size_t size"
+.Ft void *
+.Fn realloc "void *ptr, size_t size"
+.Ft void
+.Fn free "void *ptr"
+.Ft void
+.Fn malloc_revoke "void"
+.In malloc_np.h
+.Ft bool
+.Fn malloc_is_revoking "void"
+.Vt int
+.Va malloc_quarantining ;
+.Sh DESCRIPTION
+The Malloc Revocation Shim (MRS) extends a CHERI-aware malloc implemention
+with revocation support, providing heap temporal safety.
+.Nm
+wraps the underlying allocator and manages quarantine queues which are revoked
+by
+.Xr cheri_revoke 2 .
+.Pp
+.Nm
+implements all standard malloc interfaces.
+For detailed documentation of those interfaces see
+.Xr jemalloc 3 .
+It also provides partial implementations of the non-standard
+.Xr jemalloc 3
+interfaces
+.Fn mallocx ,
+.Fn rallocx ,
+.Fn dallocx ,
+and
+.Fn sdallocx .
+These differ from the underlying implementations in that they always zero
+the returned allocation, ignoring passed flags.
+The undocumented
+.Xr jemalloc 3
+3.0 interfaces are not implemented by
+.Nm
+because no CheriABI program should have ever used them.
+.Pp
+The
+.Fn malloc_revoke
+function triggers a full revocation pass, returning when the pass is
+complete.
+This results in the revocation of all pointers staged for quarantine by
+.Fn free
+or
+.Fn realloc .
+.Sh ENABLING REVOCATION
+Because heap temporal safety is an experimental feature with performance
+impacts that are not fully characterized,
+.Nm
+provides a number of policy mechanism to control enabling quarantine making
+use of
+.Xr cheri_revoke 2 .
+When a program using
+.Nm
+starts up, an array of parameters are evaulated to determine if free'd
+objects should be quarantined and revoked or if the underlying
+implementations should be exposed directly.
+MRS considers the following factors in order of incresing precidence:
+.Bl -bullet
+.It
+The value of the
+.Dv security.cheri.runtime_quarantine_default
+sysctl value.
+This is exposed by the kernel via the presence or absence of the
+.Dv ELF_BSDF_CHERI_REVOKE
+flag and the absense of the
+.Dv ELF_BSDF_CHERI_REVOKE_FORCED
+flag in the
+.Dv AT_BSDFLAGS
+ELF auxiliary argument variable.
+.It
+The value of the
+.Va malloc_quarantining
+variable if it is
+.Dv MQ_ENABLED
+or
+.Dv MQ_DISABLED .
+.Va malloc_quarantining
+is a weak symbol in libc which can be provided by the program to over
+ride the system default or prevent administrative overrides via
```

Comment:
```suggestion
is a weak symbol in libc which can be provided by the program to
override the system default or prevent administrative overrides via
```

Reply:
It is somewhat odd to mention procctl and elfctl flags here, but I appreciate you want to fully define the variable the first time you mention it.  I wonder if it would be useful to have one list that defines the various mechanisms first, and then a second list that gives the order of precedence.  That might end up awfully wordy though.

Reply:
I've prototyped a rewrite in a separate commit. Arguably the AT_BSDFLAGS bits belong somewhere else than mrs.3, but for now putting everything in one place seems to make sense.